Disable slicing when unavailable

eyal0 opened the pull request:
Disable the slicing button when slicing isn't possible.

Slicing is possible only when there is at least one slicer and that slicer is either not running on the octoprint device or we are neither printing nor paused.

Fixes kennethjiang/OctoPrint-Slicer#17

kennethjiang commented:
thank you for submitting this PR. I'm a little tied up this week. will get to it as soon as I get a chance!

kennethjiang commented on octoprint_slicer/static/js/slicer.js:
```diff
@@ -499,6 +506,42 @@ $(function() {
+	}
+	self.slicingViewModel.slicers.subscribe(self.updateSlicerProperties);
+
+	self.sameDevice = ko.computed(function() {
```
Sorry again for the delay!

What does `sameDevice` do? It looks like we don't need the new api `/slicer/properties` if we can somehow get away without `sameDevice`.

Other parts of the PR looks great. Awesome job @eyal0

eyal0 replied:
Thanks.  I will explain.

The code to allow slicing is here:

https://github.com/foosel/OctoPrint/blob/master/src/octoprint/server/api/files.py#L532

If the slicer is running on the same device as octoprint, slicing is disabled while printing so that the slicing will not use up all the CPU and ruin the print.  But if the slicer is on a different machine (maybe the actual slicing is on a different computer) then slicing is possible even while printing.

In the SlicingViewModel, is_printing and is_paused are available but "same_device" is not.  So there needs to be a way for OctoPrint to ask the server which slicers are same device and which aren't.  That is the Blueprint that I added.

It would be better if the response were available from here:

https://github.com/foosel/OctoPrint/blob/master/src/octoprint/server/api/slicing.py#L72

Then we wouldn't need to add an API and there would be less network traffic.  I will ask @foosel if she can add it to the list.  Let's wait with this PR until we get an answer.

kennethjiang replied:
Obviously you know a lot more about inner working of OctoPrint than I do @eyal0 !

I might have misunderstood what `same_device` in OctoPrint is, but I estimate for most OctoPrint users (like me) `same_device` is always `true`. 

If so, I'd propose a 4th option: drop `sameDevice` from this PR. For most users, it'll still function perfectly. For the few users whose `same_device` is some times `false`, they won't be able to slice a model when they are printing on a different Raspberry Pi, but it's probably not end of the world for them.

When your PR #1748 is accepted, we can just add a tiny bit of code to make even that use case go away.

The reason why I propose this option is to see if there is a way to use only 20% of the code to make 80% of the use cases work. But please let me know if I'm wrong on the "20%" part. For example, if you will often run into this problem (not being able to slice when you should), my proposal won't fly.

And yes there is a way to specify the minimum version of OctoPrint this plugin requires.

eyal0 replied:
Just a lot of grep to find the same_device.

For most people, I get that same_device is True, like you said.  So I'll modify this PR so that we just assume same_device is True for all users.  If @foosel later accepts https://github.com/foosel/OctoPrint/issues/1748 , we can add the support for same_device.  That'll be easy.

I'll start working on the new PR.

kennethjiang replied:
Thank you for taking my proposal!

You don't have to create a new PR. I believe if you push new commits to your branch `disable_slicing_when_unavailable` the existing PR will get updated.

eyal0 commented:
Okay, I sent a PR to @foosel.  If it is accepted then the API that I added is unneeded.  In that case, there are choices:

1. Remove the new API that I added.  That would be the cleanest solution but we would need a fallback in case the version of OctoPrint is too old to support.  Fallback could just be to default "true" or "false".

2. Retain the new API that I added.  We could take from the slicing API when it's available and only make the extra call to the new API is sameDevice isn't in the properties.

3. something else?  Is there a way to make it so that new OctoPrint-Slicer requires new OctoPrint?  Maybe in the __plugin_check__?  Kind of a pain to maintain that, though, and it's a problem for users that don't want to upgrade.

Not sure which is best.

eyal0 commented:
Okay, should be fine now, without the same_device support.  I don't have a 3d printer at the moment to test the functionality so either test it for me, thanks, or wait until next week and I can test it.  Up to you.
